### PR TITLE
[ JPG - 006] Create Transaction: CREDIT and DEBIT

### PR DESCRIPTION
**Pull request by Bhavesh-SP18**
**Create Transaction: CREDIT**

### CODE REVIEW

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    private String note;
+
+    @Enumerated(EnumType.STRING)
+    private TransactionType transactionType;
+
+    private String message;
+
+    private Float closingBalance;
+
+    @JsonFormat(pattern = "MMM dd, yyyy hh:mm:ss a", timezone = "Europe/Zagreb")
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java:34]: "Europe/Zagreb" whats this?

File: src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.UUID;
+
+@Data
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+@Entity
+@Table(name = "transaction", schema = "public")
+public class TransactionEntity {
+    @Id
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java:35]: Same mistake again

File: src/main/java/com/passbook/sparkeighteen/service/UserService.java [hunk 3]
```diff
@@ -54,6 +54,7 @@ public LoginResponse login(final LoginRequest userLogin) throws Exception {
 
         return LoginResponse.builder()
                 .userID(user.getId())
+                .firstname(user.getFirstname())
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/UserService.java:57]: This not part of this PR remove it.

File: src/main/resources/application.properties [hunk 4]
```diff
@@ -1,7 +1,7 @@
 spring.datasource.url=jdbc:postgresql://localhost:5432/passbook
 spring.datasource.username=postgres
 spring.datasource.password=qwerty1201
-#spring.jpa.hibernate.ddl-auto=create-drop
+#spring.jpa.hibernate.ddl-auto=create
```
Comment (spark-coders) [inline src/main/resources/application.properties:4]: Why are you always committing this again and again.

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+@RestController
+public class TransactionController {
+    private final TransactionService transactionService;
+
+    public TransactionController(TransactionService transactionService) {
+        this.transactionService = transactionService;
+    }
+
+    @ApiOperation("user can make Transaction")
+    @PostMapping("/{userID}/deposit")
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java:28]: How to you handle both credit and debit think about it.

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.validation.Valid;
+
+@RequestMapping("/api/v1/user")
+@RestController
+public class TransactionController {
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java:20]: How will i see the transaction of that user ?

File: src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    private UUID referralNumber;
+
+    @Column(name = "type")
+    @Enumerated(EnumType.STRING)
+    private TransactionType transactionType;
+
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "user_id", nullable = false)
+    @OnDelete(action = OnDeleteAction.CASCADE)
+    private UserEntity user;
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java:66]: Please read about mapped by ?

File: src/main/java/com/passbook/sparkeighteen/peristence/repository/UserRepository.java [hunk 8]
```diff
@@ -4,9 +4,12 @@
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.stereotype.Repository;
 
+import java.math.BigInteger;
 import java.util.Optional;
 
 @Repository
 public interface UserRepository  extends JpaRepository<UserEntity, Integer> {
     Optional<UserEntity> findByEmail(String email);
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/peristence/repository/UserRepository.java:12]: Enter missing

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        if (request.getAmount() <= 0) {
+            return Optional.empty();
+        }
+
+        TransactionEntity transaction = transactionRepository.save(TransactionEntity.builder().amount(request.getAmount()).note(request.getNote()).user(user).closingBalance(request.getAmount() + getZeroOrLastBalance(user)).time(LocalDateTime.now()).transactionType(TransactionType.CREDIT).build());
+
+        return Optional.of(transaction);
+    }
+
+    private Float getZeroOrLastBalance(UserEntity user) {
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:66]: Why this method?

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+                .amount(transaction.getAmount())
+                .note(transaction.getNote())
+                .time(transaction.getTime())
+                .closingBalance(transaction.getClosingBalance())
+                .transactionType(transaction.getTransactionType())
+                .message(String.format("Your A/C XXXXX has a %s by Rs %.2f on %s ", transaction.getTransactionType(), transaction.getAmount(), transaction.getTime()))
+                .build();
+    }
+
+    private Optional<TransactionEntity> deposit(UserEntity user, TransactionRequest request) {
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:56]: Why this method?

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    private final TransactionRepository transactionRepository;
+    private final UserRepository userRepository;
+
+    public TransactionService(TransactionRepository transactionRepository, UserRepository userRepository) {
+        this.transactionRepository = transactionRepository;
+        this.userRepository = userRepository;
+    }
+
+    public TransactionResponse transact(Integer userID, TransactionRequest request, TransactionType txnType) throws Exception {
+
```
Comment (spark-coders) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:28]: Business  logic is over complicated think more and refactor code.

**Review by spark-coders — COMMENTED** [review]
Test cases are missing 
And java docs are also missing .

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java (hunk 5 above)
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java:28]: url is wrong

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    @Enumerated(EnumType.STRING)
+    private TransactionType transactionType;
+
+    private String message;
+
+    private Float closingBalance;
+
+    @JsonFormat(pattern = "MMM dd, yyyy hh:mm:ss a", timezone = "Asia/Kolkata")
+    @NotNull
+    private LocalDateTime time;
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java:36]: variable name makes sense?

File: src/main/java/com/passbook/sparkeighteen/peristence/repository/TransactionRepository.java [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+import java.math.BigInteger;
+import java.util.List;
+import java.util.Optional;
+
+@Repository
+public interface TransactionRepository extends JpaRepository<TransactionEntity, Integer> {
+    Optional<TransactionEntity> findById(BigInteger integer);
+
+    Optional<List<TransactionEntity>> findByUser(UserEntity user);
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/peristence/repository/TransactionRepository.java:16]: You don't need to wrap a collection in Optional. Collection will be empty if there are no results found.

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        Optional<UserEntity> optionalUser = userRepository.findById(userID);
+        if (optionalUser.isEmpty())
+            return TransactionResponse.builder()
+                    .message("User not found. Please use a registered user")
+                    .build();
+
+        UserEntity user = optionalUser.get();
+
+        Optional<TransactionEntity> optionalTxn = transaction(user, request);
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:37]: why?

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+        if (optionalTxn.isEmpty())
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:39]: why?

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+                    .message("Deposit amount cannot be 0 or less.")
+                    .build();
+
+        TransactionEntity transaction = optionalTxn.get();
+
+        return TransactionResponse.builder()
+                .amount(transaction.getAmount())
+                .note(transaction.getNote())
+                .time(transaction.getTime())
+                .closingBalance(transaction.getClosingBalance())
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:50]: formatting issue

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+                .transactionType(transaction.getTransactionType())
+                .message(String.format("Your A/C XXXXX has a %s by Rs %.2f on %s ", transaction.getTransactionType(), transaction.getAmount(), transaction.getTime()))
+                .build();
+    }
+
+    private Optional<TransactionEntity> transaction(UserEntity user, TransactionRequest request) {
```
Comment (kunalcodess) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:56]: what is this method doing?

**Review by kunalcodess — CHANGES_REQUESTED** [review]

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java (hunk 1 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java (hunk 2 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/service/UserService.java (hunk 3 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/UserService.java:57]: Updated, As perviously discussed need this to be closed

File: src/main/resources/application.properties (hunk 4 above)
Comment (Ketan-SP18) [reply src/main/resources/application.properties:4]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java (hunk 5 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java (hunk 6 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java (hunk 7 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/entity/TransactionEntity.java]: Read, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/peristence/repository/UserRepository.java (hunk 8 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/repository/UserRepository.java:12]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 9 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 10 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 11 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:44]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java (hunk 5 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/controller/TransactionController.java]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java (hunk 12 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/POJO/TransactionResponse.java:40]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/peristence/repository/TransactionRepository.java (hunk 13 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/repository/TransactionRepository.java]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 14 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 15 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 16 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:92]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 17 above)
Comment (Ketan-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java]: Updated, As perviously discussed need this to be closed

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 18]
```diff
@@ -26,31 +29,86 @@ public TransactionService(TransactionRepository transactionRepository, UserRepos
 
     /**
      * Create method to do the actual transaction contaning the business logic to add deposit and withdraw and return response according
+     *
```
Comment (Aashishverma06) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java]: Remove This.

File: src/test/java/com/passbook/sparkeighteen/controller/TransactionControllerTest.java [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+import com.passbook.sparkeighteen.service.TransactionService;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+
```
Comment (Aashishverma06) [inline src/test/java/com/passbook/sparkeighteen/controller/TransactionControllerTest.java:15]: Extra space.

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 20]
```diff
@@ -2,11 +2,14 @@
 
 import com.passbook.sparkeighteen.peristence.POJO.TransactionRequest;
 import com.passbook.sparkeighteen.peristence.POJO.TransactionResponse;
+import com.passbook.sparkeighteen.peristence.entity.TransactionEntity;
 import com.passbook.sparkeighteen.peristence.entity.UserEntity;
 import com.passbook.sparkeighteen.peristence.repository.TransactionRepository;
 import com.passbook.sparkeighteen.peristence.repository.UserRepository;
 import org.springframework.stereotype.Service;
```
Comment (Aashishverma06) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java:10]: remove this.

File: src/main/java/com/passbook/sparkeighteen/peristence/repository/UserRepository.java [hunk 21]
```diff
@@ -4,6 +4,7 @@
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.stereotype.Repository;
```
Comment (Aashishverma06) [inline src/main/java/com/passbook/sparkeighteen/peristence/repository/UserRepository.java:6]: Remove this.

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java [hunk 22]
```diff
@@ -26,31 +29,86 @@ public TransactionService(TransactionRepository transactionRepository, UserRepos
...
+                .closingBalance(transaction.getClosingBalance())
+                .transactionType(transaction.getTransactionType())
+                .message(String.format("Your A/C XXXXX has a %s by Rs %.2f on %s ", transaction.getTransactionType(), transaction.getAmount(), transaction.getTime()))
+                .build();
     }
 
     /**
      * This method(getZeroOrLastBalance) is to get balance as zero if its user first transaction or the closing balance of the latest transaction.
      * first assign the balance 0.
+     *
```
Comment (Aashishverma06) [inline src/main/java/com/passbook/sparkeighteen/service/TransactionService.java]: Remove this.

File: src/test/java/com/passbook/sparkeighteen/controller/TransactionControllerTest.java [hunk 23]
```diff
@@ -0,0 +1,122 @@
+package com.passbook.sparkeighteen.controller;
+
```
Comment (Aashishverma06) [inline src/test/java/com/passbook/sparkeighteen/controller/TransactionControllerTest.java:2]: Remove this.

File: src/main/java/com/passbook/sparkeighteen/peristence/repository/UserRepository.java (hunk 21 above)
Comment (Bhavesh-SP18) [reply src/main/java/com/passbook/sparkeighteen/peristence/repository/UserRepository.java:6]: yes

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 22 above)
Comment (Bhavesh-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java]: ok, done.

File: src/main/java/com/passbook/sparkeighteen/service/TransactionService.java (hunk 18 above)
Comment (Bhavesh-SP18) [reply src/main/java/com/passbook/sparkeighteen/service/TransactionService.java]: yes.

File: src/test/java/com/passbook/sparkeighteen/controller/TransactionControllerTest.java [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+                .closingBalance(null)
+                .build());
+
+        ResponseEntity<TransactionResponse> response = transactionController.transact(txnID,transactionRequest);
+        assertEquals(response.getStatusCode(), HttpStatus.OK);
+
+    }
+
+    @Test
+    public void validRequest_successfulResponse_DebitTransactionSuccessfull() throws Exception {
```
Comment (Hrishikesh-SP18) [inline src/test/java/com/passbook/sparkeighteen/controller/TransactionControllerTest.java:75]: naming convention ?

File: src/test/java/com/passbook/sparkeighteen/controller/TransactionControllerTest.java [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+                .closingBalance(0f)
+                .build());
+
+        ResponseEntity<TransactionResponse> response = transactionController.transact(txnID,transactionRequest);
+        assertEquals(response.getStatusCode(), HttpStatus.OK);
+
+    }
+
+    @Test
+    public void InvalidRequest_errorResponse_DebitTransactionUnSuccessfull() throws Exception {
```
Comment (Hrishikesh-SP18) [inline src/test/java/com/passbook/sparkeighteen/controller/TransactionControllerTest.java:100]: naming convention ? (methodname_senario_result)

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java [hunk 26]
```diff
@@ -0,0 +1,220 @@
...
+    private UserRepository userRepository;
+
+    @Mock
+    private TransactionRepository transactionRepository;
+
+    @InjectMocks
+    private TransactionService transactionService;
+
+    @Test
+    public void InvalidPayload_userIdNotFound_transactUnsuccessfull() throws Exception {
```
Comment (Hrishikesh-SP18) [inline src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:39]: naming convention ? (methodname_senario_result)

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java [hunk 27]
```diff
@@ -0,0 +1,220 @@
...
+        Mockito.when(userRepository.findById(userID)).thenReturn(Optional.ofNullable(user));
+        Mockito.when(transactionRepository.save(ArgumentMatchers.any())).thenReturn(transaction);
+
+        TransactionResponse response = transactionService.transact(userID, request);
+        assertEquals(String.format("Your A/C XXXXX has a %s by Rs %.2f on %s ", transaction.getTransactionType(), transaction.getAmount(), transaction.getTime()), response.getMessage());
+
+    }
+
+    @Test
+    public void transact_amountLessThanOne_txnTypeCredit_transactUnsuccessful() throws Exception {
```
Comment (Hrishikesh-SP18) [inline src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:95]: naming convention ? (methodname_senario_result)

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java [hunk 28]
```diff
@@ -0,0 +1,220 @@
...
+        TransactionEntity txn = TransactionEntity.builder()
+                .amount(request.getAmount())
+                .build();
+
+        TransactionResponse response = transactionService.transact(userID, request);
+        assertEquals("Invalid Transaction amount", response.getMessage());
+    }
+
+    @Test
+    public void transact_Credit_TransactionSuccessful() throws Exception {
```
Comment (Hrishikesh-SP18) [inline src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:114]: naming convention ? (methodname_senario_result)

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java [hunk 29]
```diff
@@ -0,0 +1,220 @@
...
+        Mockito.when(userRepository.findById(userID)).thenReturn(Optional.ofNullable(user));
+        Mockito.when(transactionRepository.save(ArgumentMatchers.any())).thenReturn(transaction);
+
+        TransactionResponse response = transactionService.transact(userID, request);
+        assertEquals(String.format("Your A/C XXXXX has a %s by Rs %.2f on %s ", transaction.getTransactionType(), transaction.getAmount(), transaction.getTime()), response.getMessage());
+
+    }
+
+    @Test
+    public void transact_Debit_TransactionUnSuccessful() throws Exception {
```
Comment (Hrishikesh-SP18) [inline src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:152]: naming convention ? (methodname_senario_result)

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java [hunk 30]
```diff
@@ -0,0 +1,220 @@
...
+
+        Mockito.when(userRepository.findById(userID)).thenReturn(Optional.ofNullable(user));
+        Mockito.when(transactionRepository.findByUser(user)).thenReturn(Optional.empty());
+
+        TransactionResponse response = transactionService.transact(userID, request);
+        assertEquals(String.format("Insufficient balance to withdraw. Current Balance: %.2f", balance), response.getMessage());
+    }
+
+    @Test
+    public void transact_Debit_TransactionSuccessful() throws Exception {
```
Comment (Hrishikesh-SP18) [inline src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:181]: naming convention ? (methodname_senario_result)

File: src/test/java/com/passbook/sparkeighteen/controller/TransactionControllerTest.java (hunk 24 above)
Comment (Ketan-SP18) [reply src/test/java/com/passbook/sparkeighteen/controller/TransactionControllerTest.java]: It depends on individual, this was my style of naming them before which i think i am correct as i followed kunal sir.

File: src/test/java/com/passbook/sparkeighteen/controller/TransactionControllerTest.java (hunk 25 above)
Comment (Ketan-SP18) [reply src/test/java/com/passbook/sparkeighteen/controller/TransactionControllerTest.java]: Depends on individual, readable and understood by others just by taking a look.

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java (hunk 26 above)
Comment (Ketan-SP18) [reply src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java]: same here

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java (hunk 28 above)
Comment (Hrishikesh-SP18) [reply src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:127]: @Ketan-SP18 it would be like makeTransaction_Credit_TransactionSuccessful() correct me if i am wrong.

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java (hunk 27 above)
Comment (Hrishikesh-SP18) [reply src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:109]: @Ketan-SP18  [creaditTransaction_amountLessThanOne_TransactionUnsuccessful]

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java (hunk 29 above)
Comment (Hrishikesh-SP18) [reply src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:169]: @Ketan-SP18  [postTransaction_debit_transactionUnSuccessful]

**Review by Aashishverma06 — APPROVED** [review]

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java (hunk 30 above)
Comment (Hrishikesh-SP18) [reply src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:201]: @Ketan-SP18 [postTransaction_debit_transactionSuccessful]

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java (hunk 28 above)
Comment (Ketan-SP18) [reply src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:127]: It shows similar understanding

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java (hunk 29 above)
Comment (Ketan-SP18) [reply src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:169]: same here

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java (hunk 30 above)
Comment (Ketan-SP18) [reply src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:201]: same here

File: src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java (hunk 27 above)
Comment (Ketan-SP18) [reply src/test/java/com/passbook/sparkeighteen/service/TransactionServiceTest.java:109]: yes agreed, seems more acceptable but mine is not wrong either.

**Review by Hrishikesh-SP18 — APPROVED** [review]
